api/hypershift/v1beta1/hostedcluster_types.go

```diff
@@ -351,141 +351,161 @@ const (
+	// +kubebuilder:validation:XValidation:rule="self.filter(s, s.servicePublishingStrategy.type == 'NodePort' && has(s.servicePublishingStrategy.nodePort) && has(s.servicePublishingStrategy.nodePort.address) && has(s.servicePublishingStrategy.nodePort.port)).all(x, self.filter(y, y.servicePublishingStrategy.type == 'NodePort' && (has(y.servicePublishingStrategy.nodePort) && has(y.servicePublishingStrategy.nodePort.address) && y.servicePublishingStrategy.nodePort.address == x.servicePublishingStrategy.nodePort.address && has(y.servicePublishingStrategy.nodePort.port) && y.servicePublishingStrategy.nodePort.port == x.servicePublishingStrategy.nodePort.port )).size() <= 1)",message="Each nodePort publishingStrategy 'nodePort' and 'hostname', and each route publishingStrategy 'hostname' must be unique within the Services list."
+	// +immutable
 	Services []ServicePublishingStrategyMapping `json:"services"`
 
-	// PullSecret references a pull secret to be injected into the container
-	// runtime of all cluster nodes. The secret must have a key named
-	// ".dockerconfigjson" whose value is the pull secret JSON.
+	// pullSecret is a local reference to a Secret that must have a ".dockerconfigjson" key whose content must be a valid Openshift pull secret JSON.
+	// If the reference is set but none of the above requirements are met, the HostedCluster will enter a degraded state.
+	// TODO(alberto): Signal this in a condition.
+	// This pull secret will be part of every payload generated by the controllers for any NodePool of the HostedCluster
+	// and it will be injected into the container runtime of all NodePools.
+	// Changing this value will trigger a rollout for all existing NodePools in the cluster.
+	// +required
 	PullSecret corev1.LocalObjectReference `json:"pullSecret"`
 
-	// SSHKey references an SSH key to be injected into all cluster node sshd
-	// servers. The secret must have a single key "id_rsa.pub" whose value is the
-	// public part of an SSH key.
-	//
+	// sshKey is a local reference to a Secret that must have a "id_rsa.pub" key whose content must be the public part of 1..N SSH keys.
+	// If the reference is set but none of the above requirements are met, the HostedCluster will enter a degraded state.
+	// TODO(alberto): Signal this in a condition.
+	// When sshKey is set, the controllers will generate a machineConfig with the sshAuthorizedKeys https://coreos.github.io/ignition/configuration-v3_2/ populated with this value.
+	// This MachineConfig will be part of every payload generated by the controllers for any NodePool of the HostedCluster.
+	// Changing this value will trigger a rollout for all existing NodePools in the cluster.
+	// +optional
 	// +immutable
 	SSHKey corev1.LocalObjectReference `json:"sshKey"`
```

Comment:
shouldn't this be a pointer with `omitempty` if its optional?

Reply:
Doesn't have to be, if we consider this a configuration API and we want the field to be discoverable and there are no required fields within the struct

Reply:
Ideally this would be a pointer and the struct would have a required field .Name. That captures the valid semantics.
However, that would be now breaking for consumers not setting the field .name which has a valid code path.
By making it optional we at least allow the use case for serialize resources i.e. yaml to be applied without the field.